### ISSUE TITLE
BooruBot v3: Electric Boogaloo

### ISSUE THREAD

**Issue by AtoraSuunva** [AtoraSuunva/Booru-Discord#14]
Let's face it, BooruBot v2 still sucks.

So I'm rewriting the entire thing.

### **[tl;dr of changes](https://github.com/AtlasTheBot/Booru-Discord/issues/14#issuecomment-275418396)**
Or scroll down for detailed changelog.

- [x] Move over to the module system I wrote
- [x] Make booru searching a module. [Here it is.](https://github.com/AtlasTheBot/booru)
- [x] Rewrite all the commands
  - [x] Search
    - [x] Search images
    - [x] And use the blacklist
    - [x] Also support `rand`/`random` again aaaaaa
    - [x] Escape `()` in image link
    - [x] Search and return a random image each time (`order:random` support for sites that don't have it in the booru module?)
    - [x] Return only sfw images in non-nsfw channels
    - [x] Make type blacklisting work #15 
    - [x] Add reaction to make it show the next result when you search an image
  - [x] Blacklist (Channels are gone, replaced with a new setting, `topicEnable`)
  - [x] `b!bl cat` blacklists tag cat
  - [x] View/Edit Settings
  - [x] Settings actually do stuff
    - [x] Add a settings helper thing to `modules.js`
    - [x] `disableDMs`: Prevent people from DMing the bot if every shared server has this enabled
    - [x] `topicEnable`: Only allow boorubot to search in channels with `bb=true` in the topic
    - [x] `nsfwServer`: Treat every channel in a server as being nsfw
    - [x] `deprecationWarning`: Show a warning if a users tries to use an old BooruBot command
    - ~~[x] `showUpdates`: Show an update log on the first command used after an update~~ Actually seems more like a bother for me and the people using it so ¯\\\_(  )ツ_/¯
    - [x] `minScore`: Minimum score images must have
  - [x] Add `link` command, see #13
  - [x] Delete Command: 
     - [x] Check for "manage messages" perm and use it if available
     - [x] Add a reaction thing to delete images?
  - [x] Add `invite` command, for both bot and server (send to DMs only?)
  - ~~[x] Add `tutorial` maybe? (Kinda done, needs debugging)~~ i give up (at least in this version)
  - [x] Deprecation warning for old BooruBot commands
- [x] Make it case-insensitive

- [x] Move to system to fully support edits?
  - [x] Also make it delete it's message if yours is deleted

- [x] Clean up code a bit
  - [x] Fix `deprecation.js` since it's not showing warning messages?
  - [x] Make `search.js` less of an assault on the eyes
    - ~~[ ] Use [awaitReaction](https://discord.js.org/#/docs/main/stable/class/Message?scrollTo=awaitReactions) instead of weird reactionAdd thing~~ dropped because it has weird issues that cause it to not be reliable enough
  - [x] Change code to use es6 array destructuring (and concat extra params)
  - [x] Allow blacklisting of ratings (`rating:s`)

- [x] Test the hell out of it (Thanks error logs)

- [x] Move old settings to new format
  - ~~[ ] Use a proper database (?)~~ dropped because the current system works well enough and is easily extendible

**Comment by AtoraSuunva** [comment]
# Current Changelog
## External Changes
Stuff that'll affect you as a user.
*Things in between `[]` are parameters for commands, don't actually include the `[]`*
*`[this/that]` means that you use either `this` OR `that`*

#### General
* Invoker changed from `=` to `booru!` or `b!`
* `b!help` will display all commands, use `b!help [command]` for extra help on a command
* Many commands now have aliases, use `b!help [command]` to see them
  * `b!help search` will display extra help for the `search` command
* Commands now accept quotes, both `"` and `'` work, so you can use spaces in parameters
  * `b!example hello 'there world'` sees two parameters, `hello` and `there world`
  * Some commands will auto-concatenate the last params (`test` takes 2 params, if you do `b!test hello there world`, `hello` is param 1, `there world` is param 2)
* Edits are fully supported, editing your message will make the bot edit its response
* Deletes its messages if yours is deleted

#### Search
* The syntax for search is now `booru!search [site] [tag1] [tag2...]` 
  * or the shorter version `b!s [site] [tag1] [tag2...]`
  * or the even shorter `b![site] [tags...]`
* `e926.net` is now supported
* `derpibooru.org` is now supported
* Search now uses an embed to display the image, reverting to image link + post link if embeds are disabled
  * The embed title ("Post 12345") is a link to the post on the booru
  * The score + rating + tags are also shown
* By default only sfw images are posted (images rated safe)
  * nsfw (images rated questionnable/explicit) are posted only in nsfw channels
  * `nsfwServer` can be used to treat every channel in the server as nsfw
* There's 2 reactions added to the search result:
  * The first one ("delet this") edits the post to hide the image and only provide a link for the curious (or masochistic)
  * The second one edits the message with the next result for the tags used

#### Blacklist
* Blacklist no longer supports channels, use the new `topicEnable` setting instead
* Blacklist command changed, syntax is now: 
  * `b!blacklist` to view blacklist
  * `b!blacklist [tag/site] [thing]` to blacklist a tag or site (`b!blacklist tag cat`)
  * `b!whitelist [tag/site] [thing]` to whitelist something
  * `b!bl` and `b!wl` are aliases for `blacklist` and `whitelist`
  * `b!bl site [all/nsfw/sfw]` is still a thing
  * `b!wl [site/tag] all` clears the blacklist for sites/tags

#### Delete
* `b!delete` command added, use it to get BooruBot to delete the last image
  * `b!delete [X]` will delete the last X images, searching the last 50 images (limited to 3 for normal users, 10 for those with "Manage_Messages", if BooruBot has "Manage_Messages" perms, limit is raised from 10 to 50)

#### Settings
* View settings with `b!settings`
* Change settings with `b!settings [setting] [newValue]`
  * Only those with "Manage_Server" perms can change settings
* Previous settings are all gone, replaced with two new settings:
  * `disableDMs`: Disables DMing the bot if all the shared servers have this enabled
  * `topicEnable`: Only allow BooruBot to search for images in channels with `bb=true` in the topic
  * `nsfwServer`: If every channel should be treated as nsfw
  * `minScore`: Minimum score an image must have before posting
  * `deprecationWarning`: Warn if you try to use an old boorubot command

#### Link
* Create embeds from posts using `link`
* `b!link https://e621.net/post/show/1192588/ambiguous_gender-blue_eyes-blush-eeveelution-glace` does http://i.imgur.com/pxLttug.png

I probably forgot some changes whoops

**Comment by AtoraSuunva** [comment]
## Internal Changes
Things that don't really directly affect users

**MUST be run using node v7.3.0 and above, using `node --harmony modules.js`.**
This is because I'm using `async` and `await` to make randSearch much more tolerable to look at.

#### Module System
You can take the module system from this bot, I don't mind.
Just keep in mind the only docs available are the comments in the code, `modules/test.js`, and this.

* Moved to a *cool and new* module system that allows me to separate commands into different files
  * Modules are stored in the directory specified in `config.json`, modules in subdirectories are allowed
  * Modules are identified by the name specified in their config

* Checks for invokers are done automatically, supporting both multiple global invokers in `config.json`, and multiple local invokers in each module's config

* Offers a bunch of helpful methods (accessed by using `bot.modules`), such as: 
  * `.shlex(str)`: Removes the (global) invoker from a command and slices the arguments up into an array
    * `b!test hello 'there world' aaa` => `['test', 'hello', 'there world', 'aaa']`
  * A few methods to handle module un/reloading:
    * `.loadModules()`: (Re)load all modules
    * `.loadModule(moduleName)`: (Re)load a specific module
    * `.unloadModule(moduleName)`: Unload a specific module
  * `.config` to access the `config.json` file, `.reloadConfig()` to reload it
  * `.reportError(Error, errType)`: DM the owner the error message, returns an errID
  * `.settings`: A class with methods to deal with settings: 
    * `.get(id)`: Get the settings for a certain ID, creating them from `./defaultSettings.json` if needed
    * `.set(id, newVal)`: Set the settings for an ID to newVal
    * `.save(id)`: Writes the settings for an ID to a file (`./settings/${id}.json`)
    * `.saveAll()`: Saves all the settings
  * `.saveAndExit()`: Saves all the settings, then exits the bot

* Fetches the owner info automatically

#### Searching
* All the searching code has moved to a package, [booru](https://github.com/AtlasTheBot/booru)
  * Booru features:
    * Promise support
    * Site aliases (`sb` => `safebooru.org`)
    * jsonfy xml and add some common props to images

#### node modules
* Works with the latest version of Discord.js (11)
* Uses `uws` for `extra speed`
* Uses `booru` to search

**Comment by AtoraSuunva** [comment]
# tl;dr
* I changed the invoker from `=` to `b!` ~~to confuse all of you~~ because `=` conflicted with way too many bots.
* Code is prettier
* Added e926.net/deribooru.org
* Removed channel blacklisting, now you need to enable `topicEnable` and put `bb=true` in the topics of channels you want to use BooruBot's search in
* nsfw images in nsfw channel only (Unless you use `nsfwServer`)
* Pictures posted as rich embeds!
* Delete pictures with reactions or by using `b!delete`
```js
//todo: write this
```
